### PR TITLE
Refactor new tab page usage of RecyclerView

### PR DESCRIPTION
**Pull request by tapanmodh**
Resolves https://github.com/brave/brave-browser/issues/21737
Resolves https://github.com/brave/brave-browser/issues/22439
Resolves https://github.com/brave/brave-browser/issues/22434

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Testing orientation fix
1. Enable Brave News
2. Open a new tab page
3. Change tablet orientation to landscape
4. Begin scrolling through news feed
5. Open another new tab
6. Begin scrolling through feed in the new tab
7. Change orientation to portrait
8. Continue scrolling news
9. During each orientation change, the screen should fill the viewport (no gaps)

### Testing velocity fix
1. Launch Brave and enable the Brave News feed
2. Once enabled, start swiping through the news feed. scrolling velocity should be smooth

### Testing settings bar fix
1. Clean Install
2. Enable Brave News, enable show brave news
3. Scroll News feed
4. Observe news settings bar

### CODE REVIEW

**Comment by bsclifton** [comment]
Couple of quick things 😄 
- Could you add a more descriptive commit message? Since it's only one commit, it's a bit hard to tell everything that's going on. Maybe something like
    ```
    Refactor new tab page usage of RecyclerView.

    This fixes some issues introduced when adding Brave News to the new tab page.
    Specifically:
    - Fix velocity issue (stops during swipe)
    - Problems after changing orientation
    - Making sure settings bar is visible

    Resolves https://github.com/brave/brave-browser/issues/21737
    Resolves https://github.com/brave/brave-browser/issues/22439
    Resolves https://github.com/brave/brave-browser/issues/22434
    ```
- Adding the resolves keywords directly into the commit (on new lines) can be useful for linking back to issues when you do blame 🙂
- Can you add a bare bones test plan for verifying the fix? Looks like the issues themselves have a good steps to reproduce which you can copy 😄 I'll update test plan to add the one for orientation for example

**Comment by bsclifton** [comment]
Built locally and ran into an issue where the page gets stuck (you can't scroll up anymore; notice how there are stats cut off at the top)
Screenshot_20220701-101410
 
Here's how I created the situation 😄 
1. Fresh install
2. Enable News
3. Scroll up and down; verify it looks good
4. Scroll to very top
5. Turn phone to horizontal orientation and let page resize
6. Open a new tab
7. In the new tab, turn phone to vertical orientation
8. Click the [2] box (to view open tabs) and then pick the first open tab
9. First open tab was horizontal orientation and you can see it resize itself
10. After it does orientation fix, you can't scroll up anymore.

I think the problem isn't scrolling but instead that the settings menu shouldn't be showing when scroll position is 0. I think the stats I'm expecting are just being hidden and that the settings menu isn't supposed to show up until you scroll down a bit more

**Comment by bsclifton** [comment]
Gave this a rebase as it was crashing for me after the last push; tested this out and latest fix works great! 😄

**Review by bsclifton — APPROVED** [review]
Needs a proper code review (I only took a glance); but functionally works great 😄

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 1]
```diff
@@ -1316,92 +616,407 @@ public void run() {
...
+            }
+        });
+    }
+
+    private void keepPosition() {
+        if (BraveActivity.getBraveActivity() != null
+                && BraveActivity.getBraveActivity().getActivityTab() != null) {
+            Tab tab = BraveActivity.getBraveActivity().getActivityTab();
+
+            int itemPosition = (tab != null) ? SharedPreferencesManager.getInstance().readInt(
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:676]: `tab != null` is obsolete as we we cannot get inside the condition if it's null

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 2]
```diff
@@ -1316,92 +616,407 @@ public void run() {
...
+                                       BRAVE_RECYCLERVIEW_POSITION + tab.getId(), 0)
+                                             : 0;
+
+            new Handler(Looper.getMainLooper()).postDelayed(() -> {
+                if (mNtpAdapter != null && mNtpAdapter.getItemCount() > itemPosition) {
+                    RecyclerView.LayoutManager manager = mRecyclerView.getLayoutManager();
+                    if (manager instanceof LinearLayoutManager) {
+                        int offsetPosition = (tab != null)
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:684]: same here with `tab`

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 3]
```diff
@@ -1316,92 +616,407 @@ public void run() {
...
                 }
-            });
+            }
+        });
+    }
+
+    private void keepPosition() {
+        if (BraveActivity.getBraveActivity() != null
+                && BraveActivity.getBraveActivity().getActivityTab() != null) {
+            Tab tab = BraveActivity.getBraveActivity().getActivityTab();
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:674]: we could move that before the condition check and make `if (BraveActivity.getBraveActivity() != null && tab != null)`

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 4]
```diff
@@ -1316,92 +616,407 @@ public void run() {
...
-                    mParentScrollView.fullScroll(NestedScrollView.FOCUS_UP);
-                    mImageCreditLayout.setAlpha(1.0f);
-                    mOptinLayout.setVisibility(View.GONE);
+    }
+
+    private int firstNewsFeedPosition() {
+        if (mNtpAdapter != null) {
+            return mNtpAdapter.getStatsCount() + mNtpAdapter.getTopSitesCount()
+                    + mNtpAdapter.getNewContentCount() + 1;
+        } else {
```
[the remark this answers is not in the document]
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:708]: @SergeyZhukovsky i think the condition is needed as mNtpAdapter could be null so in that case, we need to return 0

[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:708]:
```
private int firstNewsFeedPosition() {
        if (mNtpAdapter != null) {
            return mNtpAdapter.getStatsCount() + mNtpAdapter.getTopSitesCount()
                    + mNtpAdapter.getNewContentCount() + 1;
        }

        return 0;
}
```

[the remark this answers is not in the document]
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:708]: Perfect. I have updated the same way. Thank you @serg

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 5]
```diff
@@ -1316,92 +615,406 @@ public void run() {
...
+                        if (itemPosition
+                                == mNtpAdapter.getStatsCount() + mNtpAdapter.getTopSitesCount()
+                                        + mNtpAdapter.getNewContentCount()) {
+                            offsetPosition -= mNtpAdapter.getTopMarginImageCredit();
+                        }
+
+                        LinearLayoutManager linearLayoutManager = (LinearLayoutManager) manager;
+                        linearLayoutManager.scrollToPositionWithOffset(
+                                itemPosition, offsetPosition);
+                        // mRecyclerView.invalidate();
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:696]: Is this a leftover?

File: android/java/res/layout/news_loading.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  android:orientation="vertical">
+
+  <ProgressBar
+    style="?android:progressBarStyleLarge"
+    android:layout_width="40dp"
+    android:layout_height="40dp"
+    android:layout_gravity="center" />
+
+</LinearLayout>
+
```
Comment (samartnik) [inline android/java/res/layout/news_loading.xml:20]: nit: last empty line

File: android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+    }
+
+    @Override
+    public int getItemCount() {
+        int statsCount = getStatsCount();
+        int topSitesCount = getTopSitesCount();
+        int newsLoadingCount = shouldDisplayNewsLoading() ? 1 : 0;
+
+        if (mIsDisplayNewsOptin) {
+            return statsCount + topSitesCount + 2 + newsLoadingCount;
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java:320]: There are lots of hardcoded `+ 2`, `+ 1` and not really obvious what they stand for. I wonder if we can give them meaningful variable names for code readability.

File: android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+            return TYPE_NEWS;
+        }
+    }
+
+    public int getStatsCount() {
+        return isStatsEnabled() ? 1 : 0;
+    }
+
+    private boolean isStatsEnabled() {
+        return true;
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java:398]: Is this for future usage here? As it's just never false.

**Review by samartnik — APPROVED** [review]

File: android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java (hunk 8 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java:398]: yes @samartnik it's for upcoming work for privacy hub. here we are exposing function. i will add a  comment above the function.

File: android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java (hunk 7 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/BraveNtpAdapter.java:320]: Great suggestion. updating it now.

**Pull request merged**